src/dune_rules/pkg_build_progress.mli

```diff
@@ -0,0 +1,14 @@
+open! Import
+
+(** An [Action_builder.t] intended to be run concurrently with the
+    build of a package that prints the build progress of that package
+    to the console. This must be enabled by the setting the
+    DUNE_CONFIG__PKG_BUILD_PROGRESS config variable to "enabled". If
+    this variable is "disabled" (the default) then this function
+    returns a no-op instead. *)
+val display_build_progress
+  :  Package.Name.t
+  -> Package_version.t
+  -> source_dir:[ `Some of Path.Build.t | `No_package_source ]
+  -> target_dir:[ `Some of Path.Build.t | `No_build_or_install_command ]
+  -> unit Action_builder.t
```

Comment:
I don't think this is right. Action_builder is for generating actions (like the name says), while progress reporting should occur when the actions are being executed.

Reply:
Fair enough, the current implementation does feel a bit hacky. Does dune have an API for registering a function to call when an action executes or is that something I'll need to add?

Reply:
No such API exists. I suppose you can just wrap existing actions with something like:
```ocaml
progn [ start_report_progress ; action ; stop_report_progress ]
```

A good first step would be to move the toolchain stuff to execution.

Reply:
Wrapping actions makes sense. I think for a first version we'd only need to print a message before starting a build action.

> A good first step would be to move the toolchain stuff to execution.

How is this related to printing progress status messages. We'll be printing status messages for all packages, not just toolchain packages.

Reply:
>  Or is there some way to make an Action.t that just calls an ocaml function when it runs?

There is. You can create a custom action that will execute whatever OCaml code you'd like. See the various `exec` functions of the custom actions that we create.

Reply:
> How is this related to printing progress status messages. We'll be printing status messages for all packages, not just toolchain packages.

If you were to add progress reporting during action execution, it would only for normal packages and not not for toolchains because toolchains are built and installed during rule generation.

However, once toolchains are installed during action execution, then progress reporting would work for both types of packages.

You can do both of these tasks in any order. I just suggested doing the simpler one first.

Reply:
Toolchains aren't built or installed during rule generation. The only unusual thing that happens with toolchains at rule generation time is the check whether the necessary toolchain package is already installed.

Reply:
Not sure how I confused those two things.

 I just realized that using `progn` for progress start/progress end will not work because it will not run termination if the action fails. I think you're better off introducing a dedicated action like:

```ocaml
val with_progress_report : Action.t -> Action.t
```

That will run the underlying action with the `Fiber.finalize` to make sure progress termination is always reported. You might need to add the action into `Dune_engine.Action` directly instead of using a custom action. Custom actions do not allow referring to other actions at the moment.

Reply:
What do you mean by "it will not run termination if the action fails"?

I don't think we need to report the completion of an action right now though. All I want for the first version of this is for dune to print a message "Downloading foo..." or "Building foo..." immediately before the download or build of a package begins.

Reply:
What can kind of progress reporting do you expect when you do run the following command:

```
$ dune build _build/_private/default/.pkg/foo/target
```

I would expect a message telling me that the build of package `foo` has started and then this message to be eliminated once the build is done. This is consistent with how we manage progress updates for regular build notifications and it would require updating the status line twice.

If you only care to print a log message, you don't need a termination action. However, that's a poorer UI and is inconsistent with the rest of dune.

Reply:
Just to be clear, if you want to write a console message as a stepping stone, that's alright. But I do think that you will need to do what I described about eventually. Otherwise, you will not have a consistent experience and it will not work well in watch mode.

Reply:
The specific problem I'm trying to solve right now is that when a user runs `dune build` in a project with a lockdir they get no output for the usually several minutes it takes for all the dependencies to be downloaded and built.

I'll have it write a console message at first since that's easy to implement and we can quickly get it in front of people to try it out.

Long term I agree that it would be nice to have a more sophisticated UI that displays the transition of each package from "downloading" to "building" to "done". This will be harder to do so I'll defer this work for a future change. Dune has the ability to render a dynamically-updating terminal UI with `Console.Status_line` but it only works for a single line so I don't see how we'd be able to show the status of all concurrently-building packages. Does dune have another UI mechanism that would be more appropriate or is that something we'd have to build?

Reply:
You can start with messages indeed.

We could extend the status line to work with more than one line. Although I think you should be able to make it work by putting all the packages being downloaded or installed in one line.

Reply:
Ok I updated this PR so that now there is an `Action.t` that prints a log message when it runs that gets injected before downloading and building a package.